django_pg_simple_hll/migrations/0001_initial.sql

```diff
@@ -0,0 +1,145 @@
+-- The state transition function
+-- it takes the current state, and the two input values
+-- hll_agg_state is the current running state, made up of bucket keys and hashes
+--  it is an array of n_buckets + 2
+-- input is the element we are considering
+-- hll_precision is the precision we are using for the approximation
+--  it use used to calculate the number of buckets
+create or replace function
+hll_bucket(hll_agg_state int[], input anyelement, hll_precision int)
+returns int[]
+language plpgsql immutable strict parallel safe as
+$$
+declare
+    n_buckets int := POW(2, hll_precision); -- eg.. for 512 buckets, hll_precision is 9 because 2**9 -> 512
+    hashed_input int := hashtext(input::text); -- transform the input into a text and hash it
+    bucket_hash int := hashed_input & ~(1 << 31); -- keep all the significant bits, without the sign
+    bucket_key int := hashed_input & (n_buckets - 1); -- bucket the hash into one of n buckets
+    -- pre fetch the hash for occupying the corresponding bucket for the input
+    -- we add 2 because
+    --  1. postgres arrays are 1-indexed
+    --  2. we have to account for an extra first element
+    current_hash int := hll_agg_state[bucket_key + 2];
+begin
+    -- we can only handle precision up to 11 or 2048 buckets
+    -- because hashtext produces a hash of int size with 31 useable bits
+    if hll_precision < 3 or hll_precision > 11 then
+        raise exception 'invalid hll_precision: % - must be between 3 and 11 inclusive', hll_precision;
+    end if;
+    -- postgres squeezes null elements in the array, we don't want that,
```

Comment:
As in it removes null values?

Reply:
it does somethign weird: if an array has null values at the beginning or the end, it will just reduce the size of the array to fit. This woud be fine if it would be at the end, but it also does it at the beginning

```
{1, Null, 3, Null, Null} -> {1, Null, 3}
{Null, Null, 1, Null, 3, Null, Null} -> {1, Null, 3}
```

So, what we do is we pre-alllocate an array of N+2 values and set the first and last element to a constant:
```
{4, Null, Null, Null, Null, 4}
```

Reply:
Is there a specific function you're calling that's doing this? I can't see any reference to this behaviour and this doesn't appear to be true just creating an array in a postgres shell:

```SQL
master@host:ukf> select '{NULL, 1, NULL, 3, NULL, NULL}';
+--------------------------------+
| ?column?                       |
|--------------------------------|
| {NULL, 1, NULL, 3, NULL, NULL} |
+--------------------------------+
SELECT 1
Time: 0.014s
```

Reply:
You are right, it has to do wih a function:
```sql
SELECT * FROM UNNEST('{NULL, NULL, 1, NULL}'::int []);
--------
 [NULL]
 [NULL]
      1
 [NULL]
(4 rows)

Time: 1.561 ms
```

with ordinality:

```sql
SELECT v, idx FROM UNNEST('{NULL, NULL, 1, NULL}'::int []) WITH ORDINALITY AS arr(v, idx);
   v    | idx
--------+-----
 [NULL] |   1
 [NULL] |   2
      1 |   3
 [NULL] |   4
(4 rows)

Time: 3.958 ms
```

and even aggregating:

```sql
SELECT ARRAY_AGG(v) FROM UNNEST('{NULL, NULL, 1, NULL}'::int []) WITH ORDINALITY AS arr(v, idx);
     array_agg
--------------------
 {NULL,NULL,1,NULL}
(1 row)

Time: 10.252 ms
```

but as soon as it is the output of a function

```sql
SELECT * FROM UNNEST(hll_bucket('{}'::int [], 'hello'::text, 6));
  unnest
-----------
 277190697
(1 row)

Time: 1.074 ms
```

or in more detail

```sql
SELECT * FROM UNNEST(hll_bucket('{}'::int [], 'hello'::text, 6)) WITH ORDINALITY AS arr(v, idx);
     v     | idx
-----------+-----
 277190697 |   1
(1 row)

Time: 9.225 ms
```

Reply:
And with the _flanking_:
```sql
SELECT * FROM hll_bucket('{}'::int[], 'hello'::text, 4);
                                          hll_bucket
----------------------------------------------------------------------------------------------
 {16,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,277190697,NULL,NULL,NULL,NULL,NULL,NULL,16}
(1 row)

Time: 3.907 ms
```

unnesting

```sql
SELECT * FROM UNNEST(hll_bucket('{}'::int[], 'hello'::text, 4)) WITH ORDINALITY AS arr(v, idx);
     v     | idx
-----------+-----
        16 |   1
    [NULL] |   2
    [NULL] |   3
    [NULL] |   4
    [NULL] |   5
    [NULL] |   6
    [NULL] |   7
    [NULL] |   8
    [NULL] |   9
    [NULL] |  10
 277190697 |  11
    [NULL] |  12
    [NULL] |  13
    [NULL] |  14
    [NULL] |  15
    [NULL] |  16
    [NULL] |  17
        16 |  18
(18 rows)

Time: 15.465 ms
```

Reply:
Slightly struggling to follow where you're saying you think the issue is. Might be clearer to see if you demo with some dummy functions. I don't think it's as simple as just "it has to do with it being the output of a function":

```SQL
master@host:ukf> create or replace function dummy_return_array() returns int [] language plpgsql immutable strict parallel safe as $$ begin return '{NULL, 1, NULL}'; end $$;
CREATE FUNCTION
Time: 0.002s
master@host:ukf> select dummy_return_array();
+--------------------+
| dummy_return_array |
|--------------------|
| [None, 1, None]    |
+--------------------+
SELECT 1
Time: 0.005s
master@host:ukf> select * from unnest(dummy_return_array()) with ordinality as arr(v, idx);
+--------+-----+
| v      | idx |
|--------+-----|
| <null> | 1   |
| 1      | 2   |
| <null> | 3   |
+--------+-----+
```

Reply:
you are right, maybe it's about the direct assignment?
```sql
CREATE OR REPLACE FUNCTION dummy_return_array(s int [])
RETURNS int []
LANGUAGE plpgsql IMMUTABLE STRICT PARALLEL SAFE AS
$$
begin
    s[3] := 1;
    s[5] := 2;
    return s;
end $$;

SELECT
    v,
    idx
FROM unnest(dummy_return_array('{}')) WITH ORDINALITY AS arr (v, idx);
CREATE FUNCTION
Time: 5.429 ms
   v    | idx
--------+-----
      1 |   1
 [NULL] |   2
      2 |   3
(3 rows)

Time: 1.637 ms```

Reply:
we could probably achieve the same thing assigning the end and the beginning of the array
```sql
CREATE OR REPLACE FUNCTION dummy_return_array(s int [], p int DEFAULT 3)
RETURNS int []
LANGUAGE plpgsql IMMUTABLE STRICT PARALLEL SAFE AS
$$
begin
    s[1] := coalesce(s[1], NULL);
    s[3] := 1;
    s[5] := 2;
    s[pow(2, p)] := coalesce(s[1], NULL);
    return s;
end $$;

SELECT
    v,
    idx
FROM unnest(dummy_return_array('{}')) WITH ORDINALITY AS arr (v, idx);
CREATE FUNCTION
Time: 11.999 ms
   v    | idx
--------+-----
 [NULL] |   1
 [NULL] |   2
      1 |   3
 [NULL] |   4
      2 |   5
 [NULL] |   6
 [NULL] |   7
 [NULL] |   8
(8 rows)

Time: 4.463 ms```

Reply:
da160e6c4b80b25596159575680ba690111cc5bf this works better